Add spec for DotGeneralOp

ghpvnist opened the pull request:
Regarding `status.md`:
* Verifier is set to `revisit` since the verifier and the XLA semantics do not account for `precision_config`.

closes #299

ghpvnist commented:
Some questions:
* I'm not sure whether the constraint that batching dimensions need to be less than contracting dimensions since this is allowed in XLA currently.
* Also not exactly sure what the `precision_config` does.
* Since verifier for `precision_config` is not present, we can have arbitrary number of precision_config values. How many precisions can the op specify?
* Currently, I've assigned type inference as `no` due to missing `precision_config` checks since it may affect type inference.

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+### Constraints
+
+  * (C1) `lhs`, `rhs` and `result` have the same element type.
+  * (C2) `lhs` and `rhs` have the same number of `batching_dimensions`,
```
size(`lhs_batching_dimensions`) $=$ size(`rhs_batching_dimensions`)  and size(`lhs_contracting_dimensions`) $=$ size(`rhs_contracting_dimensions`)

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+  * (C1) `lhs`, `rhs` and `result` have the same element type.
+  * (C2) `lhs` and `rhs` have the same number of `batching_dimensions`,
+  `contracting_dimensions`.
+  * (C3) `lhs_batching_dimensions` and `lhs_contracting_dimensions` values are
```
Just like reverse op
All dimensions in `lhs_batching_dimensions`, `lhs_contracting_dimensions` are unique

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+  unique.
+  * (C4) `rhs_batching_dimensions` and `rhs_contracting_dimensions` values are
+  unique.
+  * (C5) `dim(lhs_batching_dimensions, i)`, `dim(rhs_batching_dimensions, i)`
```
Can you elaborate what are we checking here?

`dim(lhs_batching_dimensions, i)` < `dim(lhs_contracting_dimensions, j)`

From the look it seems: The dimension size of lhs_batching_dimensions need to be less that the dimension sizes of contracting dimensions

ghpvnist replied:
This constraint should be removed.

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+  * (C5) `dim(lhs_batching_dimensions, i)`, `dim(rhs_batching_dimensions, i)`
+  $\lt$ `dim(lhs_contracting_dimensions, j)`,
+  `dim(rhs_contracting_dimensions, j)` for all `i` and for all `j`.
+  * (C6) 0 $\le$ `lhs_batching_dimensions[d]`, `lhs_contracting_dimensions[d]`
```
For C6-9:

 lhs_batching_dimensions is a 1-D tensor having one dimension (or axis) `0`, and so when we are saying  rhs_batching_dimensions[d] for all dimensions `d` it be ambiguous. How about 

0<= rhs_batching_dimensions[i] < rank(`lhs`) for all i $\in$ [0, size(`rhs_batching_dimensions`))

from https://github.com/openxla/stablehlo/issues/289#issuecomment-1273993605, we have plans to include size.

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+  $\lt$ rank(`lhs`).
+  * (C7) 0 $\le$ `rhs_batching_dimensions[d]`, `rhs_contracting_dimensions[d]`
+  $\lt$ rank(`rhs`) for all dimensions `d`.
+  * (C8) `dim(lhs, lhs_batching_dimensions[i])` =
```
same here 

From C2 we already established that sizes of lhs and rhs batching dimensions are the same. 
`dim(lhs, lhs_batching_dimensions[i])` = `dim(rhs, rhs_batching_dimensions[i])`  for all  i $\in$ [0, size(`rhs_batching_dimensions`))

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+  `dim(rhs, rhs_contracting_dimensions[i])` for all dimensions `i`.
+  * (C10) `dim(result, i)` is equal to (in order):
+    * `lhs_batching_dimensions[i]` for all 0 $\le$ `i` $\lt$
+    rank(`lhs_batching_dimensions`).
```
Did you mean size? rank(lhs_batching_dimensions) is 1

ghpvnist replied:
Thanks for the catch!

sdasgup3 commented:
>> the check that batching dimension numbers need to be less than contracting dimension numbers is not present in the verifier, but is in sync with XLA semantics

Did you mean the batching dimensions need to appear before the contracting dimensions?

sdasgup3 commented:
>> Also not exactly sure what the precision_config does.

[This](https://github.com/google/jax/blob/9bb2c999d6b7eb574c23209a2ec543abad891f7f/jax/_src/lax/lax.py#L658) gives a good explanation.

burmako commented on stablehlo/dialect/StablehloOps.td:
````diff
@@ -2048,9 +2048,24 @@ def StableHLO_DotGeneralOp: StableHLO_ShapedInterfaceOp<"dot_general",
+
+    Example:
+    ```mlir
+    %0 = "stablehlo.dot_general"(%arg0, %arg1) {
````
Let's use `%result`, `%lhs` and `%rhs` to be consistent with the Examples section above?

ghpvnist replied:
Sure. I had them named as such since the other ops in the file is following this format. I'll make another PR renaming the examples to match the spec.

burmako reviewed (changes requested):
Haven't reviewed Constraints yet, because I understand that both Semantics and Constraints are WIP. Please re-request review when you get a chance to push the update to Semantics.

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+
+### Constraints
+
+  * (C1) `lhs`, `rhs` and `result` have the same element type.
```
IMO, for StableHLO DotGeneralOp, we can consider the preferred element type to be same as the output element type.

According https://github.com/tensorflow/tensorflow/blob/9315da3fe49a8b36eadcc7e539b8692012320303/tensorflow/compiler/xla/service/shape_inference.cc#L768

should we write that the result type needs to be wider than any element type of the operand only for non floating-point type.

ghpvnist replied:
I've updated the comment regarding some doubts I had, and it makes sense to use the result element type as the `preferred_element_type`, and thanks for catching this. XLA doc does not mention the use of `preferred_element_type` either. I'm wondering if this was left out deliberately since this goes into how the hardware computes it.

ghpvnist commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+
+### Semantics
+
+Performs general dot products between vectors, vector/matrix and matrix/matrix
```
Thanks, I've updated the description. Let me know if this explains better.

ghpvnist commented on docs/spec_draft.md:
```diff
@@ -543,6 +544,124 @@ produces an implementation-defined value.
+
+| Name                    | Type                                                                                                                                                                                                 |
+|-------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `lhs`                   | tensor of integer, floating-point or complex element types                                                                                                                                           |
```
Thanks, I've updated the PR. The booleans are treated as 1s and 0s for the purpose of the dot product, which makes sense.

ghpvnist commented:
Update:

I spent a couple of days last week figuring out the execution semantics of this op and implemented the interpreter for DotGeneral #337 as a result. I spent majority of the time verifying whether the constraints I put were correct as they were different from XLA semantics. It turns out that the constraints are now the same as XLA's and I now have a better understanding of how this op works. 

The PR is now updated with a proposed formal notation for review.

ghpvnist commented:
Merging the interpreter PR #337 into this PR.

sdasgup3 commented on stablehlo/reference/Ops.cpp:
```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
+      op.getDotDimensionNumbers().getLhsContractingDimensions();
+  ArrayRef<int64_t> rhsContractingDims =
+      op.getDotDimensionNumbers().getRhsContractingDimensions();
+  std::vector<int64_t> lhsNonContractingDims;
```
rename it to lhsNonBatchingNonContractingDims? 
Same with rhsNonContractingDims?

sdasgup3 commented on stablehlo/reference/Ops.cpp:
```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
+            rhsContractingDims.end())
+      rhsNonContractingDims.push_back(i);
+  std::vector<int64_t> contractingDimSizes;
+  for (uint64_t i = 0; i < lhsContractingDims.size(); ++i)
```
fuse the two loops to collect contractingDimSizes and compute   totalContractingSize

sdasgup3 commented on stablehlo/reference/Ops.cpp:
```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
+       ++resIt) {
+    Type resElTy =
+        op->getResultTypes().front().dyn_cast<ShapedType>().getElementType();
+    Element sum;
```
Can we extract the computation to get 0-initialized sum in a lambda `initializeAccumulator`? That will make the code within the current loop more clear..

sdasgup3 commented on stablehlo/reference/Ops.cpp:
```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
+                   &roundingErr);
+      sum = Element(resElTy, std::complex<APFloat>(real, imag));
+    }
+    std::vector<int64_t> lhsIdx(lhs.getType().getRank());
```
Can you add some comment here (may be using a simple example) to explain what are we doing below?

sdasgup3 replied:
Something like for batching and non-batching-non-constracting dimensions the indices of result, lhs and rhs are the same.

sdasgup3 commented:
> Here is a suggestion for formalism:
> 
> Let batching_dims = [0, 1, .., size(`lhs_batching_dims`)]
> 
> foreach dimension pair (c,c') ∈ (lhs_contracting_dimension, rhs_contracting_dimensions)
> 
> result[...,ip,...]=∑(c,c′)=(0,0)(dim(lhs,c),dim(rhs,c′))lhs(...,jc,...,jq,...)∗rhs(...,kc′,...,kr,...)
> 
> * If p is in `batching dimension`, j[`lhs_batching_dims`[p]] = ip and k[`rhs_batching_dims`[p]] = ip
> * jq = kr = ip Otherwise // For non batching and non contracting


@burmako 
From reviewing implementation and feed back from @ghpvnist, it seems the underlying algorithm formalized here is right. But the way it is formalized might  be confusing and need re-work. wdyt?

ghpvnist commented:
@sdasgup3 @burmako Here is a proposed reformulation. Wdyt?
Let:
* `batching_dims` = [0, 1, .., size(`lhs_batching_dims`)].
* LCD = `lhs_contracting_dimensions`.
* RCD = `rhs_contracting_dimensions`.

$$ `result`[...,\ ip,\ ...] = \sum_{d=0}^{size(LCD)} \sum_{(c,\ c')=(0,\ 0)}^{(dim(lhs,\ LCD[d]),\ dim(rhs,\ RCD[d]))} lhs(...,\ jc,\ ...,\ jq,\ ...)\ *\ rhs(...,\ kc',\ ...,\ kr,\ ...)$$

* If p is in `batching_dims`, j[`lhs_batching_dims`[p]] = ip and k[`rhs_batching_dims`[p]] = ip.
* jq = kr = ip otherwise (For non batching and non contracting dimensions).

sdasgup3 commented:
> @sdasgup3 @burmako Here is a proposed reformulation. Wdyt? Let:

Looks good. 

After re-thinking it seems that we need to reword the last line with:  `jq = kr = ip otherwise` (same for the earlier proposed formulation https://github.com/openxla/stablehlo/pull/307#issuecomment-1276910301). No need to mention `For non batching and non contracting dimensions` as p in `result` cannot be a contracting dimension.

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -663,6 +664,142 @@ produces an implementation-defined value.
+
+| Name                         | Type                                         |
+|------------------------------|----------------------------------------------|
+| `lhs`                        | tensor of any supported types                |
```
type? and below

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -663,6 +664,142 @@ produces an implementation-defined value.
+| `rhs_batching_dimensions`    | 1-dimensional tensor constant of type `si64` |
+| `lhs_contracting_dimensions` | 1-dimensional tensor constant of type `si64` |
+| `rhs_contracting_dimensions` | 1-dimensional tensor constant of type `si64` |
+| `precision_config`           | constant of type `si64`                      |
```
How about 

constant of type `enum {DEFAULT, HIGHEST, HIGH}` ?

@burmako @ghpvnist 
IMO introducing enum does not need introduction. We wanted to convey that precision_config takes special constant values w/o dealing with how the enum case is [actually stored](https://github.com/llvm/llvm-project/blob/1f54641d33d57d85d925e72ffda9c76aef4e9601/mlir/include/mlir/IR/EnumAttr.td#L42).

sdasgup3 replied:
From https://github.com/openxla/stablehlo/blob/07578f02f7a6d1477e43a3c8f52db6c5ec33eb68/stablehlo/dialect/StablehloEnums.td#L46 it seems like an optional array, where we can have multiple constraints specified. Do you know what is the semantics if the the array.

In the constraints we can mention that the `precision config` is optional and what the default value is.

ghpvnist replied:
Fair point. I have no objections. I'll make the changes once everyone is in agreement.

ghpvnist replied:
Looking back at previous PRs with the new enum phrasing, I'll change this accordingly.

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -663,6 +664,142 @@ produces an implementation-defined value.
+  * (C10) `dim(lhs, lhs_contracting_dimensions[i])` =
+  `dim(rhs, rhs_contracting_dimensions[i])` for all `i` $\in$ [0,
+  size(`lhs_contracting_dimensions`)).
+  * (C11) `dim(result, i)` is equal to (in order):
```
The i in `result` is different than  index `i` of lhs and rhs

How about:

The shape of `result` is given by:

{`lhs_batching_dimensions[i]` :  for all 0 $\le$ `i` $\lt$ size(`lhs_batching_dimensions`)} + {`dim(lhs, i)` : for all i $\in$ [0, rank(`lhs`)) where `i` $\notin$ `lhs_batching_dimensions` and `i` $\notin$ `lhs_contracting_dimensions`} + {`dim(rhs, i)` : for all i $\in$ [0, rank(`rhs`)) where `i` $\notin$ `rhs_batching_dimensions` and `i` $\notin$ `rhs_contracting_dimensions`}

ghpvnist replied:
Thanks, this is much cleaner

sdasgup3 commented on docs/spec_draft.md:
```diff
@@ -664,6 +665,143 @@ produces an implementation-defined value.
+
+`precision_config` controls the tradeoff between speed and accuracy for
+computations on accelerator backends. This can be one of the following:
+
```
You can add the constraints on precision config following the  changes in 
https://github.com/openxla/stablehlo/pull/441

sdasgup3 commented:
>> Type inference is set to no since precision_config is not accounted for.
missing checks for `precision_config` should only affect the verification, right? The type inference must be `yes`.

burmako commented on docs/spec_draft.md:
```diff
@@ -941,6 +942,143 @@ produces an implementation-defined value.
+
+* If p is in `batching_dims`, j[`lhs_batching_dims`[p]] = ip and
+k[`rhs_batching_dims`[p]] = ip.
+* jq = kr = ip otherwise.
```
Proposed rewording:

```
Computes dot products between slices of `lhs` and slices of `rhs` and produces
`result`.

More formally, `result[result_index] = dot_product` where
  * `lhs_result_dimensions = [d for d in axes(lhs) and d not in lhs_batching_dimensions and d not in lhs_contracting_dimensions]`.
  * `rhs_result_dimensions = [d for d in axes(rhs) and d not in rhs_batching_dimensions and d not in rhs_contracting_dimensions]`.
  * `result_batching_index + result_lhs_index + result_rhs_index = result_index`
    where `size(result_batching_index) = size(lhs_batching_dimensions)`,
    `size(result_lhs_index) = size(lhs_result_dimensions)` and
    `size(result_rhs_index) = size(rhs_result_dimensions)`.
  * `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
  * `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ...:])`.
  * `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(lhs, lhs_contracting_dimensions))`.
  * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
  * `transposed_rhs_slice = slice(result_batching_index + result_rhs_index + [:, ...:])`.
  * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
  * `dot_product = reduce(
    inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)], 
    init_values=[0],
    dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
    body=lambda x, y: add(x, y))`.
```

Feedback:
  1. Proposed wording reformulates "general dot product" as "dot products between slices of lhs and slices of rhs", which more clearly summarizes the behavior of the op.
  2. In the last few weeks, we've been gravitating towards expressing ops via other ops if possible, and the proposed formalism goes along the same lines.

ghpvnist replied:
This can be left for future work, but does this new formulation affect the current interpreter implementation since we want to have a reference implementation for the interpreter?

burmako replied:
Let's leave resolving this to when we get to the reference implementation. Generally, it is an open question of how the spec is going to be mapped on the interpreter.

ghpvnist commented on docs/spec_draft.md:
```diff
@@ -941,6 +942,143 @@ produces an implementation-defined value.
+computations on accelerator backends. This can be one of the following:
+  * `DEFAULT`: Fastest mode, but least accurate (default value).
+  * `HIGH`: Slower, but more accurate.
+  * `HIGHEST`: Slowest, but most accurate.
```
If we avoid `DEFAULT`, what about the enum under Inputs and examples?

burmako replied:
Sorry for the confusion. What I meant is to propose removing "(default value)" from the first bullet. I agree that `DEFAULT` needs to stay, at least for now.

ghpvnist commented on stablehlo/reference/Interpreter.cpp:
```diff
@@ -87,6 +87,11 @@ llvm::Expected<SmallVector<Tensor>> eval(func::FuncOp func,
       Tensor runtimeOperand = fetchOperand(cosineOp.getOperand());
       Tensor runtimeResult = eval(cosineOp, runtimeOperand);
       populateResults({runtimeResult});
+    } else if (auto dotGeneralOp = dyn_cast<DotGeneralOp>(op)) {
```
Sure, I'll reopen the ticket and migrate the updated interpreter changes out of this PR.

ghpvnist commented on docs/spec_draft.md:
```diff
@@ -941,6 +942,143 @@ produces an implementation-defined value.
+computations on accelerator backends. This can be one of the following:
+  * `DEFAULT`: Fastest mode, but least accurate (default value).
+  * `HIGH`: Slower, but more accurate.
+  * `HIGHEST`: Slowest, but most accurate.
```
Fast vs slow ultimately depends on how much precision you want to have as the final result. If intermediate results are stored using more bits than its original element type (for example, storing f32 element on a f64 or storing f32 elements in two f16, etc.), then depending on how we implement or is implemented in hardware, more precision/calculations are needed to store the intermediate results. Slowest meaning it has the most precise result approximating the original number vs fastest has the least precise of other enum types at the cost of precision.

What about the following? Feel free to suggest something else:
```
  * `DEFAULT`: Fastest calculation, but least accurate approximation to the original number.
  * `HIGH`: Slower calculation, but more accurate approximation to the original number.
  * `HIGHEST`: Slowest calculation, but most accurate approximation to the original number.
```

ghpvnist replied:
Here is a more detailed comment on this: https://github.com/google/jax/blob/9bb2c999d6b7eb574c23209a2ec543abad891f7f/jax/_src/lax/lax.py#L661-L679

burmako replied:
Reading this without the requisite background knowledge, as a StableHLO producer I wouldn't know what exactly to expect here and as a StableHLO consumer I wouldn't know how to implement this. However, given that numerical accuracy is TBD at the moment, I think it's fine to leave things in this state. I've opened a ticket to elaborate further: https://github.com/openxla/stablehlo/issues/755.

ghpvnist commented on docs/spec_draft.md:
```diff
@@ -1928,6 +1929,123 @@ produces an implementation-defined value.
+| `rhs_batching_dimensions`    | 1-dimensional tensor constant of type `si64`                |
+| `lhs_contracting_dimensions` | 1-dimensional tensor constant of type `si64`                |
+| `rhs_contracting_dimensions` | 1-dimensional tensor constant of type `si64`                |
+| `precision_config`           | variadic number of enum of `DEFAULT`, `HIGH`, and `HIGHEST` |
```
I'm wondering whether we should use "or" here since it doesn't make sense for the two `precision_config` values to have different enum types (e.g. [DEFAULT, HIGHEST]) since this would result in opposing precision.
```
variadic number of enum of `DEFAULT`, `HIGH`, or `HIGHEST`
```

ghpvnist replied:
Also noticed that `PACKED_NIBBLE` has been added recently:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/xla_data.proto#L788

burmako replied:
Nice find! Yes, indeed it's been added not so long ago. We have https://github.com/openxla/stablehlo/issues/742 to track potentially adding that to StableHLO.

burmako commented on stablehlo/dialect/StablehloOps.td:
```diff
@@ -2231,10 +2231,25 @@ def StableHLO_DotGeneralOp: StableHLO_ShapedInterfaceOp<"dot_general",
+        rhs_contracting_dimensions = [1]
+      >,
+      precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]
+    } : (tensor<2x2x2xi32>, tensor<2x2x2xi32>) -> tensor<2x2x2xi32>
```
Let's update the .td file as well - both the first sentence and the example.

burmako commented on docs/spec_draft.md:
```diff
@@ -1928,6 +1929,123 @@ produces an implementation-defined value.
+    `size(result_lhs_index) = size(lhs_result_dimensions)` and
+    `size(result_rhs_index) = size(rhs_result_dimensions)`.
+  * `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
+  * `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ...:])`.
```
Looks like I made a typo here and missed a comma: "[:, ..., :]". Same below.

burmako reviewed (changes requested):
Two minor comments, and we're good to go.

burmako commented:
Oops sorry assigned the PR to Sandeep by mistake.
